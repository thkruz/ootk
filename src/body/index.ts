export { Celestial } from './Celestial';
export { Earth } from './Earth';
export { Moon } from './Moon';
export { Sun } from './Sun';
