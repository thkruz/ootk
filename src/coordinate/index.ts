export * from './Hill.js';
export * from './ClassicalElements.js';
export * from './EquinoctialElements.js';
export * from './FormatTle.js';
export * from './Geodetic.js';
export * from './ITRF.js';
export * from './J2000.js';
export * from './RelativeState.js';
export * from './RIC.js';
export * from './StateVector.js';
export * from './TEME.js';
export * from './Tle.js';
