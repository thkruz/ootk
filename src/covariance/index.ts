export * from './CovarianceSample.js';
export * from './StateCovariance.js';
