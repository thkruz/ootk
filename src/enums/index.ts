export { CatalogSource } from './CatalogSource.js';
export { CommLink } from './CommLink.js';
export { OrbitRegime } from './OrbitRegime.js';
export { Sgp4OpsMode } from './Sgp4OpsMode.js';
export { AngularDiameterMethod } from './AngularDiameterMethod.js';
export { AngularDistanceMethod } from './AngularDistanceMethod.js';
export { PassType } from './PassType.js';
