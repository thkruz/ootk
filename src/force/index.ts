export { EarthGravity } from './EarthGravity.js';
export { Force } from './Force.js';
export { ForceModel } from './ForceModel.js';
export { Gravity } from './Gravity.js';
export { SolarRadiationPressure } from './SolarRadiationPressure.js';
export { ThirdBodyGravity } from './ThirdBodyGravity.js';
export { Thrust } from './Thrust.js';

