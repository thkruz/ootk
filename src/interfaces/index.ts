export type { GroundPositionParams } from './GroundPositionParams.js';
export type { OptionsParams } from './OptionsParams.js';
export type { TimeVariables } from './TimeVariables.js';
export type { StarObjectParams } from './StarObjectParams.js';
export type { ClassicalElementsParams } from './ClassicalElementsParams.js';
export type { EquinoctialElementsParams } from './EquinoctialElementsParams.js';
export type { ScenarioParams } from '../objects/Scenario.js';
export type { BaseObjectParams } from '../objects/base/BaseObject.js';
export type { GroundObjectParams } from '../objects/base/GroundObject.js';
export type { FacilityParams } from '../objects/base/Facility.js';
export type { SensorParams } from '../objects/attachable/SensorAdv.js';
export type { RectangularSensorParams } from '../objects/attachable/RectangularSensor.js';
