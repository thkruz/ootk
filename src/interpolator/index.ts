export { ChebyshevCoefficients } from './ChebyshevCoefficients.js';
export { ChebyshevInterpolator } from './ChebyshevInterpolator.js';
export { CubicSpline } from './CubicSpline.js';
export { CubicSplineInterpolator } from './CubicSplineInterpolator.js';
export { Interpolator } from './Interpolator.js';
export { LagrangeInterpolator } from './LagrangeInterpolator.js';
export { StateInterpolator } from './StateInterpolator.js';
export { VerletBlendInterpolator } from './VerletBlendInterpolator.js';

