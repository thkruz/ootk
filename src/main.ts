/**
 * @author @thkruz Theodore Kruczek
 * @description Orbital Object ToolKit (ootk) is a collection of tools for working
 * with satellites and other orbital objects.
 * @license AGPL-3.0-or-later
 * @copyright (c) 2020-2024 Theodore Kruczek
 *
 * Many of the classes are based off of the work of @david-rc-dayton and his
 * Pious Squid library (https://github.com/david-rc-dayton/pious_squid).
 *
 * Orbital Object ToolKit is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * Orbital Object ToolKit is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with
 * Orbital Object ToolKit. If not, see <http://www.gnu.org/licenses/>.
 */

export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './time/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
export * from './operations/operations.js';
export { BaseObject, Sensor } from './objects/index.js';
export { Earth } from './body/index.js';
export * from './coordinate/index.js';
export * from './observation/index.js';
export * from './data/DataHandler.js';
export * from './sgp4/index.js';
export * from './objects/index.js';
export * from './interpolator/index.js';
export * from './force/index.js';
export * from './maneuver/index.js';
export * from './optimize/index.js';
export * from './orbit_determination/index.js';
export * from './body/index.js';
export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './coordinate/index.js';
export * from './operations/index.js';
export * from './force/index.js';
export * from './propagator/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
