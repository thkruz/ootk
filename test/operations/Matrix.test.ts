import { Matrix, Radians, Vector } from '../../src/main';

describe('Matrix', () => {
  // should create a matrix with the correct number of rows and columns
  it('should create a matrix with the correct number of rows and columns', () => {
    const elements = [
      [1, 2, 3],
      [4, 5, 6],
      [7, 8, 9],
    ];
    const matrix = new Matrix(elements);

    expect(matrix.elements).toEqual(elements);
    expect(matrix.rows).toBe(elements.length);
    expect(matrix.columns).toBe((elements[0] as number[]).length);
  });

  // should add two matrices correctly
  it('should add two matrices correctly', () => {
    const matrix1 = new Matrix([
      [1, 2, 3],
      [4, 5, 6],
      [7, 8, 9],
    ]);
    const matrix2 = new Matrix([
      [9, 8, 7],
      [6, 5, 4],
      [3, 2, 1],
    ]);
    const result = matrix1.add(matrix2);

    expect(result.elements).toEqual([
      [10, 10, 10],
      [10, 10, 10],
      [10, 10, 10],
    ]);
    expect(result.rows).toBe(3);
    expect(result.columns).toBe(3);
  });

  // should subtract two matrices correctly
  it('should subtract two matrices correctly', () => {
    const matrix1 = new Matrix([
      [1, 2, 3],
      [4, 5, 6],
      [7, 8, 9],
    ]);
    const matrix2 = new Matrix([
      [9, 8, 7],
      [6, 5, 4],
      [3, 2, 1],
    ]);
    const result = matrix1.subtract(matrix2);

    expect(result.elements).toEqual([
      [-8, -6, -4],
      [-2, 0, 2],
      [4, 6, 8],
    ]);
    expect(result.rows).toBe(3);
    expect(result.columns).toBe(3);
  });

  // should scale a matrix correctly
  it('should scale a matrix correctly', () => {
    const matrix = new Matrix([
      [1, 2, 3],
      [4, 5, 6],
      [7, 8, 9],
    ]);
    const result = matrix.scale(2);

    expect(result.elements).toEqual([
      [2, 4, 6],
      [8, 10, 12],
      [14, 16, 18],
    ]);
    expect(result.rows).toBe(3);
    expect(result.columns).toBe(3);
  });

  // should multiply two matrices correctly
  it('should multiply two matrices correctly', () => {
    const matrix1 = new Matrix([
      [1, 2, 3],
      [4, 5, 6],
      [7, 8, 9],
    ]);
    const matrix2 = new Matrix([
      [9, 8, 7],
      [6, 5, 4],
      [3, 2, 1],
    ]);
    const result = matrix1.multiply(matrix2);

    expect(result.elements).toEqual([
      [30, 24, 18],
      [84, 69, 54],
      [138, 114, 90],
    ]);
    expect(result.rows).toBe(3);
    expect(result.columns).toBe(3);
  });

  // should multiply a matrix and a vector correctly
  it('should multiply a matrix and a vector correctly', () => {
    const matrix = new Matrix([
      [1, 2, 3],
      [4, 5, 6],
      [7, 8, 9],
    ]);
    const vector = new Vector([1, 2, 3]);
    const result = matrix.multiplyVector(vector);

    expect(result.elements).toEqual([14, 32, 50]);
  });

  // should create a matrix with empty rows
  it('should create a matrix with empty rows', () => {
    const matrix = new Matrix([[], [], []]);

    expect(matrix.elements).toEqual([[], [], []]);
    expect(matrix.rows).toBe(3);
    expect(matrix.columns).toBe(0);
  });

  // should create a matrix with empty columns
  it('should create a matrix with empty columns', () => {
    const matrix = new Matrix([[1], [2], [3]]);

    expect(matrix.elements).toEqual([[1], [2], [3]]);
    expect(matrix.rows).toBe(3);
    expect(matrix.columns).toBe(1);
  });

  // should multiply a matrix and a 3D vector correctly
  it('should multiply a matrix and a 3D vector correctly when the matrix and vector are valid', () => {
    const matrixElements = [
      [1, 2, 3],
      [4, 5, 6],
      [7, 8, 9],
    ];
    const vectorElements = [1, 2, 3];
    const expectedElements = [14, 32, 50];
    const matrix = new Matrix(matrixElements);
    const vector = new Vector(vectorElements);

    const result = matrix.multiplyVector(vector);

    expect(result.elements).toEqual(expectedElements);
  });

  // should calculate the outer product of two matrices correctly
  it('should calculate the outer product of two matrices correctly', () => {
    const matrix1 = new Matrix([
      [1, 2, 3],
      [4, 5, 6],
      [7, 8, 9],
    ]);
    const matrix2 = new Matrix([
      [9, 8, 7],
      [6, 5, 4],
      [3, 2, 1],
    ]);
    const expected = new Matrix([
      [9, 16, 21],
      [24, 25, 24],
      [21, 16, 9],
    ]);
    const result = matrix1.outerProduct(matrix2);

    expect(result).toEqual(expected);
  });

  // should calculate the transpose of a matrix correctly
  it('should calculate the transpose of a matrix correctly', () => {
    const elements = [
      [1, 2, 3],
      [4, 5, 6],
      [7, 8, 9],
    ];
    const matrix = new Matrix(elements);
    const transpose = matrix.transpose();

    expect(transpose.elements).toEqual([
      [1, 4, 7],
      [2, 5, 8],
      [3, 6, 9],
    ]);
    expect(transpose.rows).toBe(matrix.columns);
    expect(transpose.columns).toBe(matrix.rows);
  });

  // should calculate the inverse of a matrix correctly
  it('should calculate the inverse of a matrix correctly', () => {
    const elements = [
      [1, 2, 3],
      [4, 5, 6],
      [7, 8, 9],
    ];
    const matrix = new Matrix(elements);
    const inverse = matrix.inverse();

    expect(inverse.elements).toMatchSnapshot();
  });

  // allZeros
  it('should return true if all elements are zero', () => {
    const matrix = Matrix.allZeros(3, 3);

    expect(matrix.elements).toEqual([
      [0, 0, 0],
      [0, 0, 0],
      [0, 0, 0],
    ]);
  });

  // fill
  it('should fill a matrix with the correct value', () => {
    const matrix = Matrix.fill(3, 3, 1);

    expect(matrix.elements).toEqual([
      [1, 1, 1],
      [1, 1, 1],
      [1, 1, 1],
    ]);
  });

  // rotX
  it('should create a rotation matrix around the x-axis', () => {
    const matrix = Matrix.rotX((Math.PI / 2) as Radians);

    expect(matrix.elements).toMatchSnapshot();
  });

  // rotY
  it('should create a rotation matrix around the y-axis', () => {
    const matrix = Matrix.rotY((Math.PI / 2) as Radians);

    expect(matrix.elements).toMatchSnapshot();
  });

  // rotZ
  it('should create a rotation matrix around the z-axis', () => {
    const matrix = Matrix.rotZ((Math.PI / 2) as Radians);

    expect(matrix.elements).toMatchSnapshot();
  });

  // identity
  it('should create an identity matrix', () => {
    const matrix = Matrix.identity(3);

    expect(matrix.elements).toMatchSnapshot();
  });

  // diagonal
  it('should create a diagonal matrix', () => {
    const matrix = Matrix.diagonal([1, 2, 3]);

    expect(matrix.elements).toMatchSnapshot();
  });

  // reciprocal
  it('should create a reciprocal matrix', () => {
    const matrix = Matrix.allZeros(3, 3).reciprocal();

    expect(matrix.elements).toMatchSnapshot();
  });

  // cholesky
  it('should create a cholesky matrix', () => {
    const matrix = Matrix.allZeros(3, 3).cholesky();

    expect(matrix.elements).toMatchSnapshot();
  });
});
