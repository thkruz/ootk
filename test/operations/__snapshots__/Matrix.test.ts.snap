// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matrix should calculate the inverse of a matrix correctly 1`] = `
Array [
  Array [
    0,
    -2.6666666666666665,
    1.6666666666666665,
  ],
  Array [
    0,
    2.333333333333333,
    -1.3333333333333333,
  ],
  Array [
    1,
    -2,
    1,
  ],
]
`;

exports[`Matrix should create a cholesky matrix 1`] = `
Array [
  Array [
    0,
    0,
    0,
  ],
  Array [
    NaN,
    NaN,
    0,
  ],
  Array [
    NaN,
    NaN,
    NaN,
  ],
]
`;

exports[`Matrix should create a diagonal matrix 1`] = `
Array [
  Array [
    1,
    0,
    0,
  ],
  Array [
    0,
    2,
    0,
  ],
  Array [
    0,
    0,
    3,
  ],
]
`;

exports[`Matrix should create a reciprocal matrix 1`] = `
Array [
  Array [
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
  ],
]
`;

exports[`Matrix should create a rotation matrix around the x-axis 1`] = `
Array [
  Array [
    1,
    0,
    0,
  ],
  Array [
    0,
    6.123233995736766e-17,
    1,
  ],
  Array [
    0,
    -1,
    6.123233995736766e-17,
  ],
]
`;

exports[`Matrix should create a rotation matrix around the y-axis 1`] = `
Array [
  Array [
    6.123233995736766e-17,
    0,
    -1,
  ],
  Array [
    0,
    1,
    0,
  ],
  Array [
    1,
    0,
    6.123233995736766e-17,
  ],
]
`;

exports[`Matrix should create a rotation matrix around the z-axis 1`] = `
Array [
  Array [
    6.123233995736766e-17,
    1,
    0,
  ],
  Array [
    -1,
    6.123233995736766e-17,
    0,
  ],
  Array [
    0,
    0,
    1,
  ],
]
`;

exports[`Matrix should create an identity matrix 1`] = `
Array [
  Array [
    1,
    0,
    0,
  ],
  Array [
    0,
    1,
    0,
  ],
  Array [
    0,
    0,
    1,
  ],
]
`;
